backend/src/main/kotlin/com/tngtech/apicenter/backend/connector/database/service/AccessRecordDatabase.kt

```diff
@@ -0,0 +1,58 @@
+package com.tngtech.apicenter.backend.connector.database.service
+
+import com.tngtech.apicenter.backend.connector.database.entity.AccessRecordEntity
+import com.tngtech.apicenter.backend.connector.database.entity.AccessRecordId
+import com.tngtech.apicenter.backend.connector.database.repository.AccessRecordRepository
+import com.tngtech.apicenter.backend.domain.entity.PermissionType
+import com.tngtech.apicenter.backend.domain.entity.ServiceId
+import com.tngtech.apicenter.backend.domain.service.PermissionsManager
+
+@org.springframework.stereotype.Service
+class AccessRecordDatabase constructor(
+        private val accessRecordRepository: AccessRecordRepository
+): PermissionsManager {
+    override fun addPermission(userId: String, serviceId: ServiceId, permission: PermissionType) {
+        val key = AccessRecordId(serviceId.id, userId)
+        val record = accessRecordRepository.findById(key)
+
+        val view: Boolean = (permission == PermissionType.VIEW) ||
+            record.map { entity -> entity.view }.orElse(false)
+
+        val viewPrereleases: Boolean = (permission == PermissionType.VIEWPRERELEASE) ||
+            record.map { entity -> entity.viewPrereleases }.orElse(false)
+
+        val edit: Boolean = (permission == PermissionType.EDIT) ||
+            record.map { entity -> entity.edit }.orElse(false)
+
+        accessRecordRepository.save(AccessRecordEntity(key, view, viewPrereleases, edit))
+    }
+
+    override fun removePermission(userId: String, serviceId: ServiceId, permission: PermissionType) {
+        val key = AccessRecordId(serviceId.id, userId)
+        val record = accessRecordRepository.findById(key)
+
+        record.ifPresent { entity ->
+            val view = if (permission == PermissionType.VIEW) false else entity.view
+            val viewPrereleases = if (permission == PermissionType.VIEWPRERELEASE) false else entity.viewPrereleases
+            val edit = if (permission == PermissionType.EDIT) false else entity.edit
+            accessRecordRepository.save(AccessRecordEntity(key, view, viewPrereleases, edit))
+        }
+    }
+
+    override fun hasPermission(userId: String, serviceId: ServiceId, permission: PermissionType): Boolean {
+        val key = AccessRecordId(serviceId.id, userId)
+        val record = accessRecordRepository.findById(key)
```

Comment:
Do we cache those results automatically?

Reply:
Sorry I don't understand, can you explain?

Reply:
Does every one of these method calls end up as a call to the database, or are they cached in between to relieve the DB?

Reply:
I don't think any caching is currently enabled https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/cache/annotation/EnableCaching.html